cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -0,0 +1,242 @@
+                   programStatusString, programRunIdString);
+        }
+      }
+
+      // Ignore notifications which specify an invalid ProgramRunId or ProgramRunStatus
+      if (programRunIdString == null || runStatus == null) {
+        return;
+      }
+
+      final ProgramRunId programRunId = GSON.fromJson(programRunIdString, ProgramRunId.class);
+      final String twillRunId = notification.getProperties().get(ProgramOptionConstants.TWILL_RUN_ID);
+      final Map<String, String> userArguments = getArguments(properties, ProgramOptionConstants.USER_OVERRIDES);
+      final Map<String, String> systemArguments = getArguments(properties, ProgramOptionConstants.SYSTEM_OVERRIDES);
+
+      final long stateChangeTime = getTime(notification.getProperties(), ProgramOptionConstants.LOGICAL_START_TIME);
+      final long endTime = getTime(notification.getProperties(), ProgramOptionConstants.END_TIME);
+      final ProgramRunStatus programRunStatus = runStatus;
+
+      switch(programRunStatus) {
+        case STARTING:
+          if (stateChangeTime == -1) {
+            throw new IllegalArgumentException("Start time was not specified in program starting notification for " +
+                                               "program run {}" + programRunId);
+          }
+          Retries.supplyWithRetries(new Supplier<Void>() {
+            @Override
+            public Void get() {
+              store.setStart(programRunId.getParent(), programRunId.getRun(),
+                             TimeUnit.MILLISECONDS.toSeconds(stateChangeTime), twillRunId,
+                             userArguments, systemArguments);
+              return null;
+            }
+          }, RetryStrategies.fixDelay(Constants.Retry.RUN_RECORD_UPDATE_RETRY_DELAY_SECS, TimeUnit.SECONDS));
+          break;
+        case RUNNING:
+          if (stateChangeTime == -1) {
+            throw new IllegalArgumentException("Run time was not specified in program running notification for " +
+                                               "program run {}" + programRunId);
+          }
+          Retries.supplyWithRetries(new Supplier<Void>() {
+            @Override
+            public Void get() {
+              store.setRunning(programRunId.getParent(), programRunId.getRun(),
+                               TimeUnit.MILLISECONDS.toSeconds(stateChangeTime), twillRunId);
+              return null;
+            }
+          }, RetryStrategies.fixDelay(Constants.Retry.RUN_RECORD_UPDATE_RETRY_DELAY_SECS, TimeUnit.SECONDS));
+          break;
+        case SUSPENDED:
+          Retries.supplyWithRetries(new Supplier<Void>() {
+            @Override
+            public Void get() {
+              store.setSuspend(programRunId.getParent(), programRunId.getRun());
+              return null;
+            }
+          }, RetryStrategies.fixDelay(Constants.Retry.RUN_RECORD_UPDATE_RETRY_DELAY_SECS, TimeUnit.SECONDS));
+          break;
+        case RESUMING:
```

Comment:
I've had to introduce `ProgramRunStatus.RESUMING` so that the service can understand a request to resume a program.

I think this is slightly better than reading the program status'es as strings, but if there is an issue with adding to this enum, then we would have to parse on the string